plugin/pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -735,3 +735,62 @@ func haveSame(a1, a2 []string) bool {
 	}
 	return false
 }
+
+type TolerationMatch struct {
+	info NodeInfo
+}
+
+func NewTolerationMatchPredicate(info NodeInfo) algorithm.FitPredicate {
+	tolerationMatch := &TolerationMatch{
+		info: info,
+	}
+	return tolerationMatch.PodToleratesNodeTaints
+}
+
+func (t *TolerationMatch) PodToleratesNodeTaints(pod *api.Pod, nodeName string, nodeInfo *schedulercache.NodeInfo) (bool, error) {
+	node, err := t.info.GetNodeInfo(nodeName)
+	if err != nil {
+		return false, err
+	}
+
+	return tolerationsToleratesTaints(pod.Spec.Tolerations, node.Status.Taints), nil
+}
+
+func tolerationsToleratesTaints(tolerations []api.Toleration, taints []api.Taint) bool {
+	// If the taint list is nil/empty, it is tolerated by all tolerations by default.
+	if len(taints) == 0 {
+		return true
+	}
+
+	// The taint list isn't nil/empty, a nil/empty toleration list can't tolerate them.
+	if len(tolerations) == 0 {
+		return false
+	}
+
+	for _, toleration := range tolerations {
```

Comment:
I think this is easier to understand if you loop over the taints instead of the tolerations (it's also compact enough to do in one function). Someting like

```
for (each taint with effect TaintEffectNoSchedule) {
    tolerated := false
    for (each toleration with effect TaintEffectNoSchedule or empty effect) {
         if (toleration.key == taint.key &&
            (toleration.operator == TolerationOpExists ||
            (toleration.operator == TolerationOpEquals && taint.value == toleration.value))) {
                  tolerated = true
                  break
          }
     }
     if (!tolerated) return false
}
return true